### PR TITLE
KAFKA-2295; Support given for dynamically loaded classes (encoders, e…

### PR DESCRIPTION
**Pull request by omkreddy**
Rebased code..

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java [hunk 1]
```diff
@@ -578,4 +578,26 @@ public static void delete(File file) {
...
+    */
+    public static ClassLoader getKafkaClassLoader() {
+        return Utils.class.getClassLoader();
+    }
+
+    /**
+     * Get the Context ClassLoader on this thread or, if not present, the ClassLoader that
+     * loaded Kafka.
+     *
+     * This should be used whenever passing a ClassLoader to Class.ForName
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: Typo, should be `Class.forName`

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java [hunk 2]
```diff
@@ -185,17 +185,14 @@ public void logUnused() {
...
         List<T> objects = new ArrayList<T>();
         for (String klass : klasses) {
-            Class<?> c;
+            Object o;
             try {
-                c = Class.forName(klass);
+                o = Utils.newInstance(klass, t);
             } catch (ClassNotFoundException e) {
-                throw new ConfigException(key, klass, "Class " + klass + " could not be found.");
+                throw new KafkaException(klass + " ClassNotFoundException exception occured");
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: Why are we swallowing the original exception?

Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: ok..added the caused exception.

**Comment by guozhangwang** [comment]
LGTM, thanks.

**Comment by b-goyal** [comment]
Rebased code..

**Pull request closed without merging**